### PR TITLE
gnrc_ipv6_nib: provide component to handle ABRO

### PR DESCRIPTION
**Pull request by miri64**
This adds a functionality to handle border router information for [multihop prefix and 6LoWPAN context dissemination](https://tools.ietf.org/html/rfc6775#section-8.1).

~~Depends on #7212.~~

This PR is part of the network layer remodelling effort:
PR dependencies

### CODE REVIEW

**Comment by miri64** [comment]
Rebased to current master and current dependencies

**Comment by miri64** [comment]
Rebased to current master

File: sys/include/net/gnrc/ipv6/nib/abr.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#define NET_GNRC_IPV6_NIB_ABR_H
+
+#include "net/ipv6/addr.h"
+#include "net/gnrc/ipv6/nib/conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if     (GNRC_IPV6_NIB_CONF_6LBR && GNRC_IPV6_NIB_CONF_MULTIHOP_P6C) || defined(DOXYGEN)
```
Comment (cgundogan) [inline sys/include/net/gnrc/ipv6/nib/abr.h]: too many white spaces for my taste :P

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c [hunk 2]
```diff
@@ -483,6 +490,133 @@ _nib_offl_entry_t *_nib_offl_iter(const _nib_offl_entry_t *last)
...
 }
 
+void _nib_pl_remove(_nib_offl_entry_t *nib_offl)
+{
+    _nib_offl_remove(nib_offl, _PL);
+#if GNRC_IPV6_NIB_CONF_MULTIHOP_P6C
+    unsigned idx = nib_offl - _dsts;
+    if (idx < GNRC_IPV6_NIB_CONF_MULTIHOP_P6C) {
+        for (_nib_abr_entry_t *abr = _abrs;
+             abr < (_abrs + GNRC_IPV6_NIB_ABR_NUMOF);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c]: similar to `_in_dsts()` you could use `_in_abrs()` here.

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c [hunk 3]
```diff
@@ -483,6 +490,133 @@ _nib_offl_entry_t *_nib_offl_iter(const _nib_offl_entry_t *last)
...
+    return abr;
+}
+
+void _nib_abr_remove(const ipv6_addr_t *addr)
+{
+    assert(addr != NULL);
+    DEBUG("nib: Removing border router %s\n", ipv6_addr_to_str(addr_str, addr,
+                                                               sizeof(addr_str)));
+    for (_nib_abr_entry_t *abr = _abrs;
+         abr < (_abrs + GNRC_IPV6_NIB_ABR_NUMOF);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c]: `_in_abrs()`?

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c [hunk 4]
```diff
@@ -483,6 +490,133 @@ _nib_offl_entry_t *_nib_offl_iter(const _nib_offl_entry_t *last)
...
+                    gnrc_sixlowpan_ctx_remove(i);
+                }
+            }
+#endif
+            memset(abr, 0, sizeof(_nib_abr_entry_t));
+        }
+    }
+}
+
+void _nib_abr_add_pfx(_nib_abr_entry_t *abr, _nib_offl_entry_t *offl)
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c]: `offl` param can be const

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c [hunk 5]
```diff
@@ -483,6 +490,133 @@ _nib_offl_entry_t *_nib_offl_iter(const _nib_offl_entry_t *last)
...
+            }
+        }
+    }
+    return NULL;
+}
+
+_nib_abr_entry_t *_nib_abr_iter(const _nib_abr_entry_t *last)
+{
+    for (const _nib_abr_entry_t *abr = (last) ? (last + 1) : _abrs;
+         abr < (_abrs + GNRC_IPV6_NIB_ABR_NUMOF);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c]: `in_abrs()`?

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h [hunk 6]
```diff
@@ -639,6 +659,65 @@ static inline void _nib_ft_remove(_nib_offl_entry_t *nib_offl)
...
+/**
+ * @brief   Adds a prefix to the managed prefix of the authoritative border
+ *          router
+ *
+ * @pre `(abr != NULL) && (offl != NULL) && (offl->mode & _PL)`
+ *
+ * @param[in] abr   The border router.
+ * @param[in] offl  The prefix to add.
+ */
+void _nib_abr_add_pfx(_nib_abr_entry_t *abr, _nib_offl_entry_t *offl);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h]: const for `offl`, maybe for `abr` also?

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h [hunk 7]
```diff
@@ -639,6 +659,65 @@ static inline void _nib_ft_remove(_nib_offl_entry_t *nib_offl)
...
+ * @pre `(abr != NULL)`
+ *
+ * @param[in] abr   The border router
+ * @param[in] last  Last prefix (NULL to start)
+ *
+ * @return  entry after @p last.
+ * @return  NULL, if @p last is the last prefix of @p abr or if @p last
+ *          wasn't in NIB (and != NULL).
+ */
+_nib_offl_entry_t *_nib_abr_iter_pfx(_nib_abr_entry_t *abr,
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h]: const for `last` ?

File: sys/net/gnrc/network_layer/ipv6/nib/nib_abr.c [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ * @file
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+
+#include "net/gnrc/ipv6/nib/abr.h"
+
+#include "_nib-internal.h"
+
+#if GNRC_IPV6_NIB_CONF_6LBR && GNRC_IPV6_NIB_CONF_MULTIHOP_P6C
+int gnrc_ipv6_nib_abr_add(ipv6_addr_t *addr)
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/nib/nib_abr.c]: const

**Comment by miri64** [comment]
First of all: rebased to current master, now addressing your comments

File: sys/include/net/gnrc/ipv6/nib/abr.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6/nib/abr.h]:
```C
#if(GNRC_IPV6_NIB_CONF_6LBR&&GNRC_IPV6_NIB_CONF_MULTIHOP_P6C)||defined(DOXYGEN)
```

better?

Comment (miri64) [reply sys/include/net/gnrc/ipv6/nib/abr.h]: (just joking, will fix)

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h (hunk 6 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h]: `abr` is manipulated by this function, `offl` can do.

**Comment by miri64** [comment]
Oops, forgot the `_in_abrs()` stuff.

**Comment by cgundogan** [comment]
@miri64 thanks for addressing my comments. It's safe to squash now!

**Comment by miri64** [comment]
Squashed.

**Comment by miri64** [comment]
Fixed and squashed immediately errors reported by Murdock.

**Comment by cgundogan** [comment]
all lights are on green!

**Pull request merged**